<a href="https://colab.research.google.com/github/phyllsmoyo/Projects/blob/branch_1/Project_Advertising_Sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### **Import Libraries**

In [5]:
#Import libraries
import pandas as pd
import numpy as np

#### **Upload the Dataset**

In [22]:
#Import the Dataset into a pandas DataFrame
!wget 'https://github.com/phyllsmoyo/Projects/blob/branch_1/Advertising%20Budget%20and%20Sales.csv'
df_data = pd.read_csv('/content/Advertising Budget and Sales.csv', index_col=0)

--2020-12-23 13:30:01--  https://github.com/phyllsmoyo/Projects/blob/branch_1/Advertising%20Budget%20and%20Sales.csv
Resolving github.com (github.com)... 140.82.112.4
Connecting to github.com (github.com)|140.82.112.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘Advertising Budget and Sales.csv.3’

Advertising Budget      [ <=>                ] 139.12K   754KB/s    in 0.2s    

2020-12-23 13:30:02 (754 KB/s) - ‘Advertising Budget and Sales.csv.3’ saved [142454]



#### **Data Preparation and Understanding**

In [11]:
#View the first 5 records
df_data.head(5)

TV Ad Budget ($)  Radio Ad Budget ($)  Newspaper Ad Budget ($)  Sales ($)
1             230.1                 37.8                     69.2       22.1
2              44.5                 39.3                     45.1       10.4
3              17.2                 45.9                     69.3        9.3
4             151.5                 41.3                     58.5       18.5
5             180.8                 10.8                     58.4       12.9

The dataset has threee features; TV Ads, Radio Ads and Newspaper Ads and the the target is Sales.

In [23]:
#View the size of the dataset
df_data.shape

(200, 4)

In [70]:
#Understand the data by viewing some basic statistical details
perc =[.20, .40, .60, .80]
df_data.describe(percentiles=perc)

TV Ad Budget ($)  ...   Sales ($)
count        200.000000  ...  200.000000
mean         147.042500  ...   14.022500
std           85.854236  ...    5.217457
min            0.700000  ...    1.600000
20%           59.180000  ...    9.700000
40%          119.000000  ...   11.800000
50%          149.750000  ...   12.900000
60%          186.060000  ...   14.800000
80%          228.540000  ...   18.920000
max          296.400000  ...   27.000000

[10 rows x 4 columns]

In [67]:
#Check if the dataset contains any null values.
df_data.isnull().any()

TV Ad Budget ($)           False
Radio Ad Budget ($)        False
Newspaper Ad Budget ($)    False
Sales ($)                  False
dtype: bool

The Dataset does not have any null values

In [15]:
#View the columns of the dataset
df_data.columns

Index(['TV Ad Budget ($)', 'Radio Ad Budget ($)', 'Newspaper Ad Budget ($)',
       'Sales ($)'],
      dtype='object')

In [25]:
#Create a feature object from the dataset
X = df_data[['TV Ad Budget ($)', 'Radio Ad Budget ($)', 'Newspaper Ad Budget ($)']]
X.head()

TV Ad Budget ($)  Radio Ad Budget ($)  Newspaper Ad Budget ($)
1             230.1                 37.8                     69.2
2              44.5                 39.3                     45.1
3              17.2                 45.9                     69.3
4             151.5                 41.3                     58.5
5             180.8                 10.8                     58.4

In [26]:
y = df_data[['Sales ($)']]
y.head()

Sales ($)
1       22.1
2       10.4
3        9.3
4       18.5
5       12.9

In [31]:
print(X.shape, y.shape, sep='\n')

(200, 3)
(200, 1)


In [33]:
#Split Test and Training sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=4)

By default the train_test_split will split 75% training and 25% test data.

In [34]:
#view the shape of the data
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape, sep='\n')

(150, 3)
(50, 3)
(150, 1)
(50, 1)


#### **Modeling Training**

In [36]:
#Import the Linear Regression Model
from sklearn.linear_model import LinearRegression

In [53]:
#Initiate the instance of the Linear Regression Model
normalize = False
linreg = LinearRegression(normalize=normalize)

In [54]:
#Train the Model
linreg.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [55]:
#Print the intercept and the coefficients
print('The Intercept is: {}'.format(linreg.intercept_))
print('The Coefficients are: {}'.format(linreg.coef_))

The Intercept is: [3.09401086]
The Coefficients are: [[0.04585077 0.17764264 0.00247669]]


In [91]:
#Prediction
y_pred = linreg.predict(X_test)
y_pred[0:5]

array([[17.21150199],
       [16.81441379],
       [21.90735108],
       [24.58034087],
       [12.22742501]])

#### **Model Evaluation**

In [92]:
#Test the Model's accuracy
#import the library
from sklearn import metrics

In [102]:
#Calculate the Metrics
#Mean Squared Error
MSE = metrics.mean_squared_error(y_test, y_pred)
#Root Mean Squared Error
RMSE = metrics.mean_squared_error(y_test, y_pred, squared=False)
#Mean Absolute Error
MAE = metrics.mean_absolute_error(y_test, y_pred)
#R-Squared
R2 = metrics.r2_score(y_test, y_pred)

In [103]:
print('Mean Squared Error: {}'.format(MSE))
print('Root Mean Squared Error: {}'.format(RMSE))
print('Mean Absolute Error: {}'.format(MAE))
print('R Squared: {}'.format(R2))

Mean Squared Error: 2.090641627700804
Root Mean Squared Error: 1.4459051240315888
Mean Absolute Error: 1.1532385280495925
R Squared: 0.9157191830180074


**R-squared** is a statistical measure of how close the data are to the fitted regression line. It is also known as the coefficient of determination, or the coefficient of multiple determination for multiple regression.

Based on R squared this means that our model is **91.572%** accurate.

Also the other metrics are small meaning that the model is doing well in predicting using unseen data.

#### **Author**

**[Sinothi Moyo](https://www.linkedin.com/in/phyllsmoyo/)**